src/compiler/checker.ts

```diff
@@ -38599,13 +38607,52 @@ namespace ts {
                     error(node, message, symbolToString(symbol));
                 }
 
-                // Don't allow to re-export something with no value side when `--isolatedModules` is set.
+                // Type assertion to defeat the type predicate aliasing of `node` here because
+                // `isTypeOnlyImportOrExportDeclaration` is a one-sided type guard so `!isDeclaredTypeOnly`
+                // does not actually imply any narrowing effect on `node`.
+                const isDeclaredTypeOnly = isTypeOnlyImportOrExportDeclaration(node as never);
                 if (compilerOptions.isolatedModules
-                    && node.kind === SyntaxKind.ExportSpecifier
-                    && !node.parent.parent.isTypeOnly
-                    && !(target.flags & SymbolFlags.Value)
-                    && !(node.flags & NodeFlags.Ambient)) {
-                    error(node, Diagnostics.Re_exporting_a_type_when_the_isolatedModules_flag_is_provided_requires_using_export_type);
+                    && !isDeclaredTypeOnly
+                    && !(node.flags & NodeFlags.Ambient)
+                    && (!(target.flags & SymbolFlags.Value) || getTypeOnlyAliasDeclaration(symbol))) {
+                    const isType = !(target.flags & SymbolFlags.Value);
+                    const typeOnlyAlias = getTypeOnlyAliasDeclaration(symbol);
+
+                    switch (node.kind) {
+                        case SyntaxKind.ImportClause:
+                        case SyntaxKind.ImportSpecifier:
+                        case SyntaxKind.ImportEqualsDeclaration: {
+                            if (compilerOptions.preserveValueImports) {
+                                const message = isType
+                                    ? Diagnostics._0_is_a_type_and_must_be_imported_with_a_type_only_import_when_preserveValueImports_and_isolatedModules_are_both_enabled
+                                    : Diagnostics._0_resolves_to_a_type_only_declaration_and_must_be_imported_with_a_type_only_import_when_preserveValueImports_and_isolatedModules_are_both_enabled;
+                                const name = idText(node.kind === SyntaxKind.ImportSpecifier ? node.propertyName || node.name : node.name!);

```

Comment:
Is there a way to structure this to avoid the postfix-`!`? `name` is optional on `ImportClause` if there's no default binding, but its unclear which part of the preceding code proves that if this is an `ImportClause`, then it has a default binding.

Reply:
I don’t think there’s a great way without restructuring a lot of calling code. If the node being checked is an `ImportClause`, it has to have a `name` in order to get to this function, or else it wouldn’t have been an alias at all (it would have been a side effect import like `import "module"`). IOW, when `ImportClause` has a symbol, it is the alias represented by its `name`.

Reply:
In that case, I would suggest a comment to that effect, or using `Debug.checkDefined(node.name, "An ImportClause with a symbol should have a name")`.